tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -7630,16 +7630,78 @@ void SortOp::getCanonicalizationPatterns(RewritePatternSet& results,
 // TransposeOp
 //===----------------------------------------------------------------------===//
 
+template <typename T>
+static DenseElementsAttr foldTransposeHelper(mhlo::TransposeOp op,
+                                             DenseElementsAttr valueAttr) {
+  llvm::SmallVector<int64_t> permutation =
+      llvm::to_vector(op.getPermutation().getValues<int64_t>());
+  int64_t rank = permutation.size();
+  auto inputShape = op.getOperand().getType().cast<ShapedType>().getShape();
+  auto outputType = op.getType().cast<ShapedType>();
+  auto outputShape = outputType.getShape();
+
+  assert(outputType.hasStaticShape() && "output shape must be static");
```

Comment:
Can you confirm (by testing with a case below) that this assertion won't fire if someone tried to specify the op as `"mhlo.transpose"(%0) {permutation = dense<[2, 1, 0]> : tensor<3xi64>} : (tensor<2x2x2xf32>) -> tensor<?x?x?xf32>` or similar (either thanks to an earlier assertion firing or type inference preventing this from becoming a problem)?

Reply:
Just to be clear on this, you don't need to add a test case, as it's just me confirming the behavior of type inference, and it would require an additional RUN: command to check the error.